### PR TITLE
adding `output.devtoolNamespace` option

### PR DESCRIPTION
**Pull request by MagicDuck**
**What kind of change does this PR introduce?**
feature

**Did you add tests for your changes?**
yes

**If relevant, link to documentation update:**

https://github.com/webpack/webpack.js.org/pull/1653

**Summary**

When loading multiple libraries built with webpack, you can run into
collisions of the sourcemap file paths. For examle, both have
"webpack:///src/index.js".

This change addresses the problem by introducing a new output option
`output.devtoolNamespace` which defaults to `output.library` when
not specified. The defaults moduleFilenameTemplates in all the
sourcemap plugins have been modified to start with:
"webpack://[namespace]/...", where [namespace] will be replaced by
the `output.devtoolNamespace`.

Notice that there are only two slashes following "webpack:" now.
This is to make it behave just as before when not building with a
namespace. When building with a namespace you only get the two
slashes, but from what I've seen the chrome dev tools only care
about the first 2 slashes anyways.

Discussed with sokra here:
https://github.com/webpack/webpack/issues/5767

**Does this PR introduce a breaking change?**

No

**Other information**

### CODE REVIEW

File: lib/ModuleFilenameHelpers.js [hunk 1]
```diff
@@ -160,3 +165,7 @@ ModuleFilenameHelpers.matchObject = function matchObject(obj, str) {
 		if(ModuleFilenameHelpers.matchPart(str, obj.exclude)) return false;
 	return true;
 };
+
+ModuleFilenameHelpers.setModuleNamespace = function setModuleNamespace(moduleNamespace) {
+	this._moduleNamespace = moduleNamespace;
```
Comment (sokra) [inline lib/ModuleFilenameHelpers.js]: Don't use static variables.

Instead pass the `namespace` to the `createFilename` method.

`createFilename(module, options, requestShortener)`

`options: { namespace, moduleFilenameTemplate }`

Also add a backward-compatibility layer when `typeof options === "string"`

File: lib/WebpackOptionsApply.js [hunk 2]
```diff
@@ -193,6 +194,12 @@ class WebpackOptionsApply extends OptionsApply {
 			ExternalsPlugin = require("./ExternalsPlugin");
 			compiler.apply(new ExternalsPlugin(options.output.libraryTarget, options.externals));
 		}
+
+		if(!options.output.devtoolNamespace) {
+			options.output.devtoolNamespace = options.output.library;
+		}
+		ModuleFilenameHelpers.setModuleNamespace(options.output.devtoolNamespace);
```
Comment (sokra) [inline lib/WebpackOptionsApply.js]: Pass `namespace` to the SourceMapDevtoolPlugin (next section) instead. Similar to `devtoolModuleFilenameTemplate`

**Comment by MagicDuck** [comment]
I'll make it default to empty string in the webpack options defaulter to pass the tests.

File: lib/EvalDevToolModulePlugin.js [hunk 3]
```diff
@@ -14,7 +14,7 @@ class EvalDevToolModulePlugin {
 
 	apply(compiler) {
 		compiler.plugin("compilation", (compilation) => {
-			compilation.moduleTemplate.apply(new EvalDevToolModuleTemplatePlugin(this.sourceUrlComment, this.moduleFilenameTemplate));
+			compilation.moduleTemplate.apply(new EvalDevToolModuleTemplatePlugin(this.sourceUrlComment, this.moduleFilenameTemplate, compiler.options.output.devtoolNamespace));
```
[the remark this answers is not in the document]
Comment (MagicDuck) [reply lib/EvalDevToolModulePlugin.js]: Do you guys have any standard for documenting what the options are?
There is no way to guess from seeing something like:
```js
	constructor(options) {
		this.options = options;
	}
```
Should I add some jsdoc comments?

[the remark this answers is not in the document]
Comment (MagicDuck) [reply lib/EvalDevToolModulePlugin.js]: or do you completely destructure all options in the constructor?

File: lib/ModuleFilenameHelpers.js [hunk 4]
```diff
@@ -48,7 +50,17 @@ function asRegExp(test) {
 	return test;
 }
 
-ModuleFilenameHelpers.createFilename = function createFilename(module, moduleFilenameTemplate, requestShortener) {
+ModuleFilenameHelpers.createFilename = function createFilename(module, options, requestShortener) {
+	let opts;
```
[the remark this answers is not in the document]
Comment (MagicDuck) [reply lib/ModuleFilenameHelpers.js]: Using `Object.assign()` can have weird consequences sometimes for your defaults:
```js
Object.assign({foo: "default"}, {foo: undefined})  --> {foo: undefined}
```
Can we use extend lib instead? (https://www.npmjs.com/package/extend)
```js
var extend = require("extend")
extend({foo: "default"}, {foo: undefined}) ---> {foo: "default"}
```

https://runkit.com/magicduck/59e8fd082fe1da001215d79f

**Comment by MagicDuck** [comment]
@sokra when you get a chance, let me know what approach I should follow regarding documenting and defaulting options (see my comments)

**Comment by sokra** [comment]
Thanks

**Comment by MagicDuck** [comment]
Yay, thanks!

**Pull request merged**